### PR TITLE
feat(broker): added bpmn element type to WorkflowInstanceRecord

### LINKED ISSUE #1839: A workflow instance record should contain the bpmn type of the element
Every workflow instance record should have a field `bpmnType` which reflects the type of the related bpmn element. The type should be an enum in the protocol listing all currently supported types:

- Start/End/Intermediate Catch/Boundary Event
- Service/Receive Task
- Sequence Flow
- Exclusive/Event Based/Parallel Gateway
- Process/Sub Process

### PR DESCRIPTION
**Pull request by miguelpires**
This PR adds a BpmnElementType to the WorkflowInstanceRecord. Couple of implementation choices that might be relevant:
* Most records should have one of the five specified bpmn element types since they will relate to a bpmn element. However, some records aren't in the scope of an element (e.g., create and cancel commands), which is why I added an UNSPECIFIED value. 
* Although not having a default element type would be good to prevent omiting future bugs, I made UNSPECIFIED the default because otherwise I'd have to add it to many classes (e.g., test rules, assert classes, broker response classes, etc) where the bpmnElementType isn't relevant. I thought the added complexity and code outweighted the advantage of not having a default.

closes #1839

### CODE REVIEW

File: broker-core/src/main/resources/io/zeebe/broker/exporter/debug/index.html [hunk 1]
```diff
@@ -161,14 +161,14 @@
             r._bpmnProcessId = r.value.bpmnProcessId;
             r._elementId = r.value.elementId;
             r._workflowInstanceKey = r.value.workflowInstanceKey;
-            r._scopeInstanceKey = r.value.scopeInstanceKey;
+            r._scopeInstanceKey = r.value.scopeInstanceKeyProp;
```
Comment (saig0) [inline broker-core/src/main/resources/io/zeebe/broker/exporter/debug/index.html]: Why did you change the name to `scopeInstanceKeyProp`?

File: broker-core/src/test/java/io/zeebe/broker/workflow/WorkflowInstanceFunctionalTest.java [hunk 2]
```diff
@@ -351,4 +355,119 @@ public void shouldReprocessWorkflowInstanceRecordsWhenWorkflowIsTemporarilyUnava
...
+            .connectTo("merge")
+            .done();
+
+    brokerRule.getClock().pinCurrentTime();
+    testClient.deploy(model);
+
+    // when
+    testClient.createWorkflowInstance(PROCESS_ID);
+
+    assertThat(RecordingExporter.jobRecords(JobIntent.CREATED).exists()).isTrue();
```
Comment (saig0) [inline broker-core/src/test/java/io/zeebe/broker/workflow/WorkflowInstanceFunctionalTest.java]: You don't need to check this here. It is already checked in `completeJobOfType()`.

File: protocol/src/main/java/io/zeebe/protocol/BpmnElementType.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.zeebe.protocol;
+
+public enum BpmnElementType {
+  UNSPECIFIED,
```
Comment (saig0) [inline protocol/src/main/java/io/zeebe/protocol/BpmnElementType.java]: I would prefer `UNKNOWN` because we already use it in similar cases.

File: protocol/src/main/java/io/zeebe/protocol/BpmnElementType.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  EVENT,
+  TASK,
+  SEQUENCE_FLOW,
+  GATEWAY,
+  PROCESS
```
Comment (saig0) [inline protocol/src/main/java/io/zeebe/protocol/BpmnElementType.java]: We need more fine granular element types. For example, to differentiate between a `startEvent`, a `intermediateCatchEvent` and a `endEvent`. A good idea could be to use the element names from the XML as element types.

**Review by saig0 — CHANGES_REQUESTED** [review]
Instead of setting the element type in each handler, you could also determine the type while transforming the element. So the handler can use the type (e.g. `element.getBpmnType()`).

File: broker-core/src/main/java/io/zeebe/broker/workflow/processor/CatchEventBehavior.java [hunk 5]
```diff
@@ -134,7 +134,9 @@ public boolean occurEventForElement(
       workflowInstanceRecord.wrap(source);
       workflowInstanceRecord.setPayload(eventPayload);
       workflowInstanceRecord.setElementId(eventHandlerId);
-      workflowInstanceRecord.setBpmnElementType(BpmnElementType.EVENT);
+      if (!source.getElementId().equals(eventHandlerId)) {
+        workflowInstanceRecord.setBpmnElementType(BpmnElementType.BOUNDARY_EVENT);
```
Comment (saig0) [inline broker-core/src/main/java/io/zeebe/broker/workflow/processor/CatchEventBehavior.java]: We should add at least a comment here in which case we need to change the element type, and when not. Otherwise, we could also lookup for the element and get the type of it.

File: broker-core/src/main/java/io/zeebe/broker/workflow/processor/gateway/ExclusiveSplitHandler.java [hunk 6]
```diff
@@ -44,7 +43,7 @@ public void handle(BpmnStepContext<ExecutableExclusiveGateway> context) {
 
       if (sequenceFlow != null) {
         value.setElementId(sequenceFlow.getId());
-        value.setBpmnElementType(BpmnElementType.SEQUENCE_FLOW);
+        value.setBpmnElementType(sequenceFlow.getElementType());
```
Comment (saig0) [inline broker-core/src/main/java/io/zeebe/broker/workflow/processor/gateway/ExclusiveSplitHandler.java]: Instead of setting the element type and element id in each handler, we could change the method `appendNewEvent()` to 
`context.getOutput().appendNewEvent(WorkflowInstanceIntent.SEQUENCE_FLOW_TAKEN, value, sequenceFlow);` So we can not forget to set the element type.

**Review by saig0 — CHANGES_REQUESTED** [review]

**Comment by menski** [comment]
@saig0 please have a look at the latest changes

**Review by saig0 — APPROVED** [review]
:+1:
